src/recipe/recipe.service.ts

```diff
@@ -1,26 +1,154 @@
+  ) {}
+
+  async createRecipe(
+    userId: number,
+    createRecipeDto: CreateRecipeDto,
+  ): Promise<Recipe> {
+    const { title, description, ingredients } = createRecipeDto;
+
+    const user = await this.userRepository.findOne({ where: { id: userId } });
+    if (!user) {
+      throw new NotFoundException('User not found');
+    }
+
+    const existingRecipe = await this.recipeRepository.createQueryBuilder('recipe').where('recipe.user = :userId',{userId:user.id}).andWhere('(recipe.title=:title OR recipe.description=:description)',{title,description}).getOne()
+    
+    if (existingRecipe) {
+      throw new ConflictException(
+        'Recipe with similar details already exists for this user',
+      );
+    }
+
+    const recipe = this.recipeRepository.create({
+      title,
+      description,
+      user,
+    });
+
+    const savedRecipe = await this.recipeRepository.save(recipe);
+
+    //check the ingredients if the ingredient not exist, create one
+    for (const ingredientData of ingredients) {
+      let ingredient = await this.ingredientRepository.findOne({
+        where: {
+          id: ingredientData.ingredientId,
+          name: ingredientData.name,
+        },
+      });
+      if (!ingredient) {
+        ingredient = this.ingredientRepository.create({
+          name: ingredientData.name,
+        });
+        ingredient = await this.ingredientRepository.save(ingredient);
+      }
+
+      const recipeIngredient = this.recipeIngredientRepository.create({
+        recipe: savedRecipe,
+        ingredient,
+        amount: ingredientData.amount,
+        unit: ingredientData.unit,
+      });
+
+      await this.recipeIngredientRepository.save(recipeIngredient);
+    }
+
+    return await this.recipeRepository.findOne({
+      where: { id: savedRecipe.id }
+    });
   }
 
   findAll() {
-    return `This action returns all recipe`;
+    return this.recipeRepository.find();
+  }
+
+  findOne(recipeId: number) {
+    return this.recipeRepository.findOne({ where: { id: recipeId } });
   }
 
-  findOne(id: number) {
-    return `This action returns a #${id} recipe`;
+  async findRecipeById(recipeId: number): Promise<Recipe> {
+    const recipe = await this.recipeRepository.findOne({
+      where: { id: recipeId },
+      relations: ['user'],
+    });
+    if (!recipe) {
+      throw new NotFoundException('Recipe not found');
+    }
+    return recipe;
   }
 
-  update(id: number, updateRecipeDto: UpdateRecipeDto) {
-    return `This action updates a #${id} recipe`;
+  async updateRecipe(
+    userId: number,
+    recipeId: number,
+    updateRecipeDto: UpdateRecipeDto,
+  ): Promise<Recipe> {
+    const recipe = await this.recipeRepository.findOne({where:{userId:userId,id:recipeId}});
+    if (!recipe) {
+      throw new ForbiddenException(`You are not allowed to update this recipe`);
+    }
+
+    Object.assign(recipe, updateRecipeDto);
+
+    if (updateRecipeDto.ingredients && updateRecipeDto.ingredients.length > 0) {
```

Comment:
can you explain why you need to delete before update recipe?

Reply:
```suggestion
    if (updateRecipeDto.ingredients && updateRecipeDto.ingredients.length > 0) {
```

Reply:
Well ,you have two options, just update the new ingredients, meaning you need to search all of the ingredinets related to the recipe and for each one of them you need to check if there has been a change and update those ingredients,or delete all of the ingredients and create new ones instad.I guess its the same comlexity. Am I wrong?